shared/crypto/slip/slip.go

```diff
@@ -0,0 +1,147 @@
+package slip
+
+import (
+	"crypto/hmac"
+	"crypto/sha512"
+	"encoding/binary"
+	"fmt"
+	"github.com/pokt-network/pocket/shared/crypto"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+const (
+	// PoktAccountPathFormat used for HD key derivation where 635 is the SLIP-0044 coin type
+	// To be used with fmt.Sprintf to generate child keys
+	// Ref: https://github.com/satoshilabs/slips/blob/master/slip-0044.md
+	PoktAccountPathFormat = "m/44'/635'/%d'" // m/purpose'/coin_type'/account_idx'
+	// As defined in: https://github.com/satoshilabs/slips/blob/master/slip-0010.md#master-key-generation
+	seedModifier = "ed25519 seed"
+	// Hardened key values, as defined in: https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki#extended-keys
+	firstHardenedKeyIndex = uint32(1 << 31) // 2^31; [0, 2^31) are non-hardened keys which are not supported for ed25519
+	maxHardenedKeyIndex   = ^uint32(0)      // 2^32-1
+	maxChildKeyIndex      = maxHardenedKeyIndex - firstHardenedKeyIndex
+)
+
+var (
+	ErrNoDerivation = fmt.Errorf("no derivation for a hardened ed25519 key is possible")
+	ErrInvalidPath  = fmt.Errorf("invalid BIP-44 derivation path")
+	pathRegex       = regexp.MustCompile(`m(/\d+')+$`)
+)
+
+type slipKey struct {
+	SecretKey []byte
+	ChainCode []byte
+}
+
+// Derives a master key from the seed provided and returns the child at the correct index
+func DeriveChild(path string, seed []byte) (crypto.KeyPair, error) {
+	// Break down path into uint32 segments
+	segments, err := pathToSegments(path)
+	if err != nil {
+		return nil, err
+	}
+
+	// Initialise a master key from seed to start child generation
+	key, err := newMasterKey(seed)
+	if err != nil {
+		return nil, err
+	}
+
+	// Iterate over segments in path regenerating the child key until correct
+	for _, i32 := range segments {
+		// Force hardened keys
```

Comment:
Does this also implicitly check that `purpose` and `coin_type` are < `maxChildKeyIndex`? If so, let's just update the comment to make it clear.